File: Java/Data Structures & Algorithms Module - Java/Week8/Revision/Week7/Hi.java
/*
Hi.java
Billy Cussen
18/11/2020
*/
public class Hi{

    public static void main(String[]args){

        long startTime = System.nanoTime();
        System.out.println("Count Hi: "+countHi("Hihihihellohellohihi"));
        long endTime = System.nanoTime();
        System.out.println("Looped Approach took "+(endTime-startTime)+" nanoseconds");
        
        startTime = System.nanoTime();
        System.out.println("Count Hi Recursion: "+countHiRecursion("Hihihihellohellohihi"));
        endTime = System.nanoTime();
        System.out.println("Recursive Approach took "+(endTime-startTime)+" nanoseconds");
    }

    public static int countHi(String word){
        word = word.toLowerCase();
        int count = 0;
        for(int i = 0; i < word.length(); i++){
            if(word.charAt(i)=='h' && word.charAt(i+1)=='i'){
                count++;
                i++;
            }
        }
        return count;
    }

    public static int countHiRecursion(String word){
        if(word.length()<=1){
            return 0;
        }
        word = word.toLowerCase();
        if(word.charAt(0)=='h'&&word.charAt(1)=='i'){
            return 1 + countHiRecursion(word.substring(2,word.length()));
        } else{
            return 0 + countHiRecursion(word.substring(1,word.length()));
        }
    }

}

File: Python/Data Structures & Algorithms Module - Python/Week7/Hi.py
"""
Hi.py
Billy Cussen
18/11/2020
"""
import timeit

def countHi(word):
    count = 0
    word = word.lower()
    for i in range(len(word)):
        if word[i] == 'h' and word[i+1]=='i':
            count+=1
            i+=1
    return count


def countHiRecursion(word):
    if len(word) <= 1:
        return 0

    if (word[0] == 'h' or word[0] == 'H') and (word[1] == 'i' or word[1] == 'I'):
        return 1 + countHiRecursion(word[2:])
    else:
        return countHiRecursion(word[1:])


startTime = timeit.default_timer()
print("Count Hi: "+str(countHi("HihiHelloHelloHihi")))
endTime = timeit.default_timer() 
nanoTime = round((endTime-startTime)*1000000000,2)
print("Looped Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer() 
print("Count Hi Recursion: "+str(countHi("HihiHelloHelloHihi")))
endTime = timeit.default_timer() 
nanoTime = round((endTime-startTime)*1000000000,2)
print("Recursive Approach took "+str(nanoTime)+" nanoseconds")